### ISSUE TITLE
[BUG] offset mismatched in gpt_dataset.py _query_document_sample_shuffle_indices

### ISSUE THREAD

**Issue by zixianwang2022** [NVIDIA/Megatron-LM#1145]
**Describe the bug**
Hi, line 276 of [megatron/core/datasets/gpt_dataset.py](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/gpt_dataset.py#L276) has an offset mismatched bug, found in the most recent commit [703cc88](https://github.com/NVIDIA/Megatron-LM/commit/703cc88a87a3ccfa84e14ae575932cbfa3fa9b7c). 

The line is:  
```
offset = 0 if i > doc_index_beg else doc_index_beg_offset
```
This is to provide the beginning offset of the next input data of the dataset. The intention should be when index `i` is at 0, the starting offset should be whatever the sample's given [doc_index_beg_offset](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/gpt_dataset.py#L247C24-L247C44). 

However, by using `offset = doc_index_beg_offset` for the index 0 example causes the input_ids of that example is loaded **only a portion** of the next data's input_ids. 

**How I found the bug**: 
I created a dummy dataset where each text begins with `Q: __question_for_the_data__. \n A: ` Every data begins with "Q: ", this will give me input_ids starting with `877, 251541` when using GPTSentencePieceTokenizer. However, when I print out the input_ids for each data in [forward_step function](https://github.com/NVIDIA/Megatron-LM/blob/ssm/pretrain_mamba.py#L167), I observed that 1 data's input_ids does not start with `877, 251541`. When I looked into the code, I found that it was the `offset` that is causing the problem. 

**To Reproduce**
Follow what I described above by creating a dummy dataset where each text begins with "Q: ". Include print statements are the two files mentioned above. I am running `Megatron-LM/examples/mamba/train.sh` with 800M model, pure ssm structure and no attention, on 1 nproc. 

**Expected behavior**
The first data will not begin with `877, 251541`. 

**Stack trace/logs**
Here is the print statement results when including print in [line 290](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/gpt_dataset.py#L290). 
```
        length = sum(map(len, sample_parts))
        from megatron.training import get_tokenizer
        tokenizer = get_tokenizer ()
        a = tokenizer.detokenize (sample_parts[0].tolist())
        print (f'\n input_ids : a: {sample_parts[0]} \n')
        print (f'\n detokenized : a: {a} \n')
        b = tokenizer.detokenize (sample_parts[1].tolist())
        print (f'\n input_ids : b: {sample_parts[1]} \n')
        print (f'\n detokenized : b: {b} \n')
        c = tokenizer.detokenize (sample_parts[2].tolist())
        print (f'\n input_ids : c: {sample_parts[2]} \n')
        print (f'\n detokenized : c: {c} \n')
```
print output: 
Theoretically, they all should print out "Q: ...". However, the first data did not begin with offset 0, which should be the start of the sentence, but rather starting from the middle. 
```
 input_ids : a: [   625   2732  43607 251514    305   2245  14389 251597     14 251519
 251541] 


 detokenized : a: which young performer, and American actress?
A: 


 input_ids : b: [   877 251541    624    431   2867   5126    276   1194   8180  69342
    291    276   1349 251514   3466    406    514    684   4325    291
    261   1632    300  90683  22886    296    580    375   4940   4600
   1012   4398  79772 251526    291    276   4750    305  14150    300
   7236   3879    300    276   8440  66333     14 251519 251541] 


 detokenized : b: Q: who was once considered the best kick boxer in the world, however he has been involved in a number of controversies relating to his "unsportsmanlike conducts" in the sport and crimes of violence outside of the ring.?
A: 


 input_ids : c: [   877 251541    796   2245   5090 251514   2204 251514    305   7179
  10956 251490    624    676   6949    291    261   2619   1632    300
 105584 251514   3872 110748    291    276 251490 251525 251561 251564
 251561   2204    375  46038   1705    305   4303  33716   1371 251597
     14 251519 251541] 


 detokenized : c: Q: what American stage, film, and television actor  who also appeared in a large number of musicals, played Samson in the 1949 film "Samson and Delilah".?
A: 
```

**Environment (please complete the following information):**
 - Megatron-LM commit ID: 703cc88
 - PyTorch version: `>>> torch.__version__.  '2.2.0a0+81ea7a4'` 
 - CUDA version
```
root@66bbe53c15b5:/workspace/megatron/examples/mamba# nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Wed_Nov_22_10:17:15_PST_2023
Cuda compilation tools, release 12.3, V12.3.107
Build cuda_12.3.r12.3/compiler.33567101_0
```  
 - NCCL version `2.8.10`

**Proposed fix**
Let `offset = 0` fixes the issue above. 

**Additional context**

**Comment by jkamalu** [comment]
This is not a bug. It is expected. For GPT pretraining, we pack samples using as many documents as is required. So when a document doesn't have enough tokens to fill a sample, we pull from other documents. This consequently means that certain documents will be part of more than one sample. In this case, the first few tokens which you're seeing "lost" likely belong to another sample in the dataset, i.e. the last few tokens at the last sample part in a different sample at a different index. Note also that you are referring to index 0 as the first sample, which is incorrect. Index 0 is remapped to another index using the shuffle index, so we're actually talking about a sample != 0 in this example.

**Comment by zixianwang2022** [comment]
> This is not a bug. It is expected. For GPT pretraining, we pack samples using as many documents as is required. So when a document doesn't have enough tokens to fill a sample, we pull from other documents. This consequently means that certain documents will be part of more than one sample. In this case, the first few tokens which you're seeing "lost" likely belong to another sample in the dataset, i.e. the last few tokens at the last sample part in a different sample at a different index. Note also that you are referring to index 0 as the first sample, which is incorrect. Index 0 is remapped to another index using the shuffle index, so we're actually talking about a sample != 0 in this example.

Hi John, thanks for the clarification! Adding on to this, I have been wanting to finetune the Mamba2 8B model using Megatron-LM. Is there a recommended to to do this besides modifying the pretrain_mamba.py? 

I don't think there's a finetune version for Mamba specifically. As a result, I have to modify the pretrain_mamba.py and even upto the dataloader to give me 1 data at a time instead many data compacted together.

**Comment by jkamalu** [comment]
The pretrain scripts are meant to serve as good examples of what's possible with megatron, but they don't cover all use cases. If the only difference between what is possible with the pretrain script and what you desire is the dataloader changing such that each dataset sample corresponds to one raw sample, then you should be able to edit the dataset class to do this fairly easily.

**Comment by jaredcasper** [comment]
Closing this issue with the explanation above, please reopen or create a new one if there are further issues.
